src/iceberg/table_metadata.cc

```diff
@@ -196,4 +198,222 @@ Status TableMetadataUtil::Write(FileIO& io, const std::string& location,
   return io.WriteFile(location, json_string);
 }
 
+// TableMetadataBuilder implementation
+
+struct TableMetadataBuilder::Impl {
+  // Base metadata (if building from existing metadata)
+  const TableMetadata* base{nullptr};
+
+  // Mutable fields that will be used to build the final TableMetadata
+  int8_t format_version;
+  std::string table_uuid;
+  std::string location;
+  int64_t last_sequence_number{TableMetadata::kInitialSequenceNumber};
+  TimePointMs last_updated_ms{std::chrono::milliseconds(0)};
+  int32_t last_column_id{0};
+  std::vector<std::shared_ptr<Schema>> schemas;
+  std::optional<int32_t> current_schema_id;
+  std::vector<std::shared_ptr<PartitionSpec>> partition_specs;
+  int32_t default_spec_id{0};
```

Comment:
Do we have predefined constants for this and below? For example, `PartitionSpec::kInitialSpecId`?